## Using cuSpatial to Correlate Taxi Data after a Format Change
In 2017, the NYC Taxi data switched from giving their pickup and drop off locations in `lat/lon` to one of 262 `LocationID`s.  While these `LocationID`s made it easier to determine some regions and borough information that was lacking in the previous datasets, it made it difficult to compare datasets before and after this transition.  

By using cuSpatial `Points in Polygon` (PIP), we can now correlate the pick up and drop of latitudes and longitudes in a reasonable timeframe.  In this notebook, we will show you how to do so.  Currently, PIP only works on 32 polygons at once, so we will show how to process this larger 263 polygon shapefile with minimal memory impact.  Estimates are that it will take approximately 3-4x longer than if we didn't have that limitation, so something to look forward to!

You may need a 16GB card or larger.

## Imports

In [1]:
import cuspatial
import geopandas as gpd
import cudf
from numba import cuda
import numpy as np

## Download Data
We're going to download each January dataset months from 2015-2018. as well as the boundary.  The following cell, when uncommented will download just over 5GB of data.  If left uncommented, it will be just under 3GB.

In [2]:
!if [ ! -f "zones.zip" ]; then curl https://s3.amazonaws.com/nyc-tlc/misc/taxi_zones.zip -o zones.zip; else echo "zones.zip found"; fi
!if [ !  -f "taxi2015.csv" ]; then curl https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_2015-01.csv -o taxi2015.csv; else echo "taxi2015.csv found"; fi
#!if [ !  -f "taxi2016.csv" ]; then curl https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_2016-01.csv -o taxi2016.csv; else echo "taxi2016.csv found"; fi   
#!if [ !  -f "taxi2017.csv" ]; then curl https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_2017-01.csv -o taxi2017.csv; else echo "taxi2017.csv found"; fi
!if [ !  -f "taxi2018.csv" ]; then curl https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_2018-01.csv -o taxi2018.csv; else echo "taxi2018.csv found"; fi 
!if [ ! -f "taxi_zone_lookup.csv" ]; then curl https://s3.amazonaws.com/nyc-tlc/misc/taxi+_zone_lookup.csv -o taxi_zone_lookup.csv; else echo "taxi_zone_lookup.csv found"; fi

zones.zip found
taxi2015.csv found
taxi2016.csv found
taxi2017.csv found
taxi2018.csv found
taxi_zone_lookup.csv found


## Read in the data
let's read in the spatial shape file and the taxi data.  While we won't use the look up data, we have it here for your reference

In [3]:
tzones = gpd.read_file('https://data.cityofnewyork.us/api/geospatial/d3c5-ddgc?method=export&format=GeoJSON')
tzones.to_file('cu_taxi_zones.shp')
taxi_zones = cuspatial.read_polygon_shapefile('cu_taxi_zones.shp')

In [4]:
taxi2015 = cudf.read_csv("taxi2015.csv")
taxi2018 = cudf.read_csv("taxi2018.csv")
taxi_zone_lookup = cudf.read_csv("taxi_zone_lookup.csv")

In [22]:
taxi2015.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2015-01-15 19:05:39   2015-01-15 19:23:42                1   
1         1  2015-01-10 20:33:38   2015-01-10 20:53:28                1   
2         1  2015-01-10 20:33:38   2015-01-10 20:43:41                1   
3         1  2015-01-10 20:33:39   2015-01-10 20:35:31                1   
4         1  2015-01-10 20:33:39   2015-01-10 20:52:58                1   

   trip_distance  pickup_longitude  pickup_latitude  RateCodeID  \
0           1.59        -73.993896        40.750111           1   
1           3.30        -74.001648        40.724243           1   
2           1.80        -73.963341        40.802788           1   
3           0.50        -74.009087        40.713818           1   
4           3.00        -73.971176        40.762428           1   

  store_and_fwd_flag  dropoff_longitude  ...  payment_type  fare_amount  \
0                  N         -73.974785  ...             1         12.0   
1                  N         -73.994415  ...             1         14.5   
2                  N         -73.951820  ...             2          9.5   
3                  N         -74.004326  ...             2          3.5   
4                  N         -74.004181  ...             2         15.0   

   extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0    1.0      0.5        3.25           0.0                    0.3   
1    0.5      0.5        2.00           0.0                    0.3   
2    0.5      0.5        0.00           0.0                    0.3   
3    0.5      0.5        0.00           0.0                    0.3   
4    0.5      0.5        0.00           0.0                    0.3   

   total_amount  PULocationID  DOLocationID  
0         17.05           185           185  
1         17.80           212           212  
2         10.80           166           166  
3          4.80           230           230  
4         16.30           162           162  

[5 rows x 21 columns]

In [23]:
taxi2018.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2018-01-01 00:21:05   2018-01-01 00:24:23                1   
1         1  2018-01-01 00:44:55   2018-01-01 01:03:05                1   
2         1  2018-01-01 00:08:26   2018-01-01 00:14:21                2   
3         1  2018-01-01 00:20:22   2018-01-01 00:52:51                1   
4         1  2018-01-01 00:09:18   2018-01-01 00:27:06                2   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            0.5           1                  N            41            24   
1            2.7           1                  N           239           140   
2            0.8           1                  N           262           141   
3           10.2           1                  N           140           257   
4            2.5           1                  N           246           239   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          4.5    0.5      0.5        0.00           0.0   
1             2         14.0    0.5      0.5        0.00           0.0   
2             1          6.0    0.5      0.5        1.00           0.0   
3             2         33.5    0.5      0.5        0.00           0.0   
4             1         12.5    0.5      0.5        2.75           0.0   

   improvement_surcharge  total_amount  
0                    0.3          5.80  
1                    0.3         15.30  
2                    0.3          8.30  
3                    0.3         34.80  
4                    0.3         16.55

In [6]:
print(taxi_zones)

(0        1
1       34
2       35
3       36
4       37
      ... 
258    349
259    350
260    351
261    352
262    354
Name: f_pos, Length: 263, dtype: int32, 0        232
1       1113
2       1121
3       1137
4       1143
       ...  
349    97738
350    98010
351    98080
352    98162
353    98192
Name: r_pos, Length: 354, dtype: int32,                x          y
0     -74.184453  40.694996
1     -74.185632  40.691648
2     -74.185911  40.691450
3     -74.186370  40.691189
4     -74.187234  40.690595
...          ...        ...
98187 -73.937646  40.781001
98188 -73.937799  40.781031
98189 -73.937794  40.780781
98190 -73.937901  40.780750
98191 -73.938046  40.780830

[98192 rows x 2 columns])


## Correllating with cuSpatial Points In Polygon
Looking at the taxi zones and the taxi2015 data, you can see that
- 12.7 million pickup locations
- 12.7 million dropoff locations
- 263 LocationID features
- 354 LocationID rings
- 98,192 LocationID coordinates

Now that we've collected the set of pickup locations and dropoff locations, we can use `cuSpatial.point_in_polygon_bitmap` to quickly determine which pickups and dropoffs occur in each borough. That is, 353 LocationID rings composed of a total of 263 LocationID features.

To do this in a memory efficient way, instead of creating two massive 12.7 million x 263 arrays, we're going to use the 32 polygon limit to our advantage and map the resulting true values in the array a new `PULocationID` and `DOLocationID`, matching the 2018 schema.  Two things to note:
1. we had to go in a reversed order for this to work.  
1. locations outside of the `LocationID` areas are `264` and `265`.  We'll be using 264 to indicate our out-of-bounds zones as no guidance was given on how to decide between the two.

In [7]:
%%time
taxi2015['PULocationID'] = 264
taxi2015['DOLocationID'] = 264
for i in reversed(range(0,len(taxi_zones[0]), 32)):
    print(i, i+32)
    pickups = cuspatial.point_in_polygon_bitmap(taxi2015['pickup_longitude'] , taxi2015['pickup_latitude'], taxi_zones[0][i:i+32], taxi_zones[1], taxi_zones[2]['x'], taxi_zones[2]['y'])
    dropoffs = cuspatial.point_in_polygon_bitmap(taxi2015['dropoff_longitude'] , taxi2015['dropoff_latitude'], taxi_zones[0][i:i+32], taxi_zones[1], taxi_zones[2]['x'], taxi_zones[2]['y'])
    for j in range(i, i+32):
        try:
            taxi2015['PULocationID'][pickups[j]] = j
        except:
            pass
        try:
            taxi2015['DOLocationID'][pickups[j]] = j
        except:
            pass

256 288
224 256
192 224
160 192
128 160
96 128
64 96
32 64
0 32
CPU times: user 3min 27s, sys: 842 ms, total: 3min 28s
Wall time: 3min 28s


In [8]:
del pickups
del dropoffs

In [9]:
taxi2015['PULocationID'].value_counts()

237    467173
160    446675
236    441389
73     437823
229    429313
        ...  
109         3
26          2
45          1
104         1
203         1
Name: PULocationID, Length: 252, dtype: int32

## Verify
Let's test our outputs with a few `LocationID` values.

In [34]:
taxi2015.query('DOLocationID == 262')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  \
18               2  2015-01-15 19:05:41   2015-01-15 19:20:36   
43               2  2015-01-25 00:13:06   2015-01-25 00:24:51   
122              1  2015-01-15 10:26:14   2015-01-15 10:42:16   
126              1  2015-01-15 10:26:15   2015-01-15 10:34:40   
138              2  2015-01-07 14:58:08   2015-01-07 15:06:36   
...            ...                  ...                   ...   
12748825         1  2015-01-10 18:16:49   2015-01-10 18:26:59   
12748826         1  2015-01-10 18:16:49   2015-01-10 18:36:50   
12748858         2  2015-01-26 06:45:01   2015-01-26 07:04:05   
12748934         2  2015-01-09 15:04:27   2015-01-09 15:07:08   
12748965         1  2015-01-10 19:01:40   2015-01-10 19:20:34   

          passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
18                      1           2.37        -73.945541        40.779423   
43                      1           3.37        -73.945511        40.773724   
122                     1           1.90        -73.949486        40.772072   
126                     1           1.30        -73.947639        40.775143   
138                     1           0.98        -73.948692        40.777782   
...                   ...            ...               ...              ...   
12748825                4           0.90        -73.946709        40.772472   
12748826                2           3.40        -73.947174        40.775917   
12748858                1           4.35        -73.946953        40.779190   
12748934                1           0.44        -73.949150        40.773232   
12748965                1           3.50        -73.950005        40.772049   

          RateCodeID store_and_fwd_flag  dropoff_longitude  ...  fare_amount  \
18                 1                  N         -73.980850  ...         11.5   
43                 1                  N         -73.987434  ...         12.5   
122                1                  N         -73.970726  ...         12.5   
126                1                  N         -73.960693  ...          7.5   
138                1                  N         -73.955284  ...          7.0   
...              ...                ...                ...  ...          ...   
12748825           1                  N         -73.958710  ...          7.5   
12748826           1                  N         -73.988899  ...         15.5   
12748858           1                  N         -73.990387  ...         15.5   
12748934           1                  N         -73.953262  ...          4.0   
12748965           1                  N         -73.991814  ...         15.0   

          extra  mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
18          1.0      0.5         0.0           0.0                    0.3   
43          0.5      0.5         0.0           0.0                    0.3   
122         0.0      0.5         0.0           0.0                    0.3   
126         0.0      0.5         1.0           0.0                    0.3   
138         0.0      0.5         1.4           0.0                    0.3   
...         ...      ...         ...           ...                    ...   
12748825    0.0      0.5         0.0           0.0                    0.3   
12748826    0.0      0.5         1.5           0.0                    0.3   
12748858    0.0      0.5         0.0           0.0                    0.3   
12748934    0.0      0.5         0.8           0.0                    0.3   
12748965    0.0      0.5         0.0           0.0                    0.3   

          total_amount  PULocationID  DOLocationID  RatecodeID  
18                13.3           262           262           1  
43                13.8           262           262           1  
122               13.3           262           262           1  
126                9.3           262           262           1  
138                9.2           262           262           1  
... 

In [35]:
taxi2015.query('PULocationID == 2')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  \
261024           1  2015-01-23 21:42:46   2015-01-23 21:46:58   
706444           1  2015-01-11 01:05:10   2015-01-11 01:20:38   
752791           2  2015-01-19 00:13:59   2015-01-19 00:28:32   
1030339          2  2015-01-18 14:37:39   2015-01-18 15:58:39   
2902909          1  2015-01-01 00:34:42   2015-01-01 00:41:22   
3499198          1  2015-01-26 06:16:15   2015-01-26 06:34:06   
3599012          2  2015-01-01 10:49:41   2015-01-01 11:16:53   
3627154          2  2015-01-06 21:06:20   2015-01-06 21:15:19   
4148598          2  2015-01-24 04:25:56   2015-01-24 04:41:15   
5004763          1  2015-01-08 12:46:12   2015-01-08 13:09:41   
5340190          2  2015-01-25 17:25:15   2015-01-25 17:25:50   
5340191          2  2015-01-25 17:25:15   2015-01-25 17:25:50   
5439288          1  2015-01-01 02:08:33   2015-01-01 02:28:12   
5772266          2  2015-01-01 04:34:15   2015-01-01 04:49:04   
5858291          1  2015-01-05 10:35:48   2015-01-05 11:16:31   
6348963          2  2015-01-04 21:35:33   2015-01-04 21:47:54   
6452048          2  2015-01-30 14:29:44   2015-01-30 14:48:53   
8537277          2  2015-01-30 20:46:46   2015-01-30 21:02:16   
8782955          2  2015-01-01 22:40:03   2015-01-01 23:11:33   
9904922          2  2015-01-10 05:25:48   2015-01-10 05:35:56   
10078590         2  2015-01-14 22:21:39   2015-01-14 22:31:19   
10209288         2  2015-01-01 02:34:57   2015-01-01 02:39:02   
10428087         2  2015-01-13 03:30:26   2015-01-13 03:39:56   
10678598         2  2015-01-21 21:07:47   2015-01-21 21:13:46   
10965732         1  2015-01-08 07:46:12   2015-01-08 07:48:52   
11089962         2  2015-01-18 10:21:52   2015-01-18 10:27:24   
11921624         2  2015-01-30 02:20:56   2015-01-30 02:38:22   
11924673         2  2015-01-25 03:37:57   2015-01-25 03:44:23   
12104255         2  2015-01-10 03:52:56   2015-01-10 03:56:31   
12253546         1  2015-01-13 12:44:21   2015-01-13 12:54:53   

          passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
261024                  2           1.00        -73.834206        40.863976   
706444                  1           5.70        -73.848969        40.868534   
752791                  1           4.57        -73.857903        40.869251   
1030339                 3          12.53        -73.845070        40.862801   
2902909                 2           2.70        -73.858749        40.865551   
3499198                 1           4.70        -73.857101        40.869263   
3599012                 2          17.43        -73.843048        40.865501   
3627154                 1           2.26        -73.833138        40.863449   
4148598                 1           2.93        -73.859879        40.871281   
5004763                 0          10.90        -73.847313        40.865318   
5340190                 1           0.00        -73.855667        40.858910   
5340191                 1           0.00        -73.855667        40.858910   
5439288                 1           7.70        -73.857132        40.869549   
5772266                 3           6.63        -73.835258        40.861816   
5858291                 1          14.10        -73.837158        40.859615   
6348963                 2           4.44        -73.854851        40.867683   
6452048                 1           3.73        -73.833405        40.859039   
8537277                 1           4.37        -73.834068        40.863255   
8782955                 1          18.48        -73.837723        40.865086   
9904922                 6           3.02        -73.853958        40.873035   
10078590                3           1.59        -73.845406        40.868031   
10209288                2           2.77        -73.834274        40.864147   
10428087                1           4.17        -73.829948        40.861145   
10678598                1           1.66        -73.842117        40.868263   
10965732           

In [36]:
taxi2018.query('PULocationID == 2')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
1231084         1  2018-01-06 12:56:22   2018-01-06 13:12:52                1   
1773053         1  2018-01-08 14:58:22   2018-01-08 15:05:58                1   
1903847         1  2018-01-08 23:13:57   2018-01-08 23:45:34                1   
5372386         1  2018-01-20 15:02:38   2018-01-20 16:10:51                1   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
1231084            7.2           1                  N             2   
1773053            3.2           1                  N             2   
1903847           12.1           1                  N             2   
5372386           17.7           2                  N             2   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
1231084            95             2         22.5    0.0      0.5        0.00   
1773053           124             1         11.0    0.0      0.5        2.35   
1903847            62             1         36.0    0.5      0.5        3.00   
5372386           162             1         52.0    0.0      0.5       11.70   

         tolls_amount  improvement_surcharge  total_amount  
1231084          0.00                    0.3         23.30  
1773053          0.00                    0.3         14.15  
1903847          0.00                    0.3         40.30  
5372386          5.76                    0.3         70.26

Sticking with `PULocationID == 2` to test that things are all working, we'll now get some numbers and use them for our next step

In [12]:
taxi2015['PULocationID'].count()

12748986

In [13]:
taxi2018['PULocationID'].count()

8759874

## Bringing Them All Together
If you wanted to include this as part of a larger clean up of Taxi data, you'd then concatenate this dataframe into a `dask_cudf` dataframe and delete its `cuDF` version, or convert it into arrow memory format and process it similar to how we did in the mortgage notebook.  For now, as we are only working on a couple of GBs, we'll concatonate in cuDF.

Before we do, let's change the `RateCodeID` in 2015 to match the `RatecodeID` column in 2018

In [25]:
taxi2015['RatecodeID'] = taxi2015['RateCodeID']
taxi2015.drop('RateCodeID')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  \
0                2  2015-01-15 19:05:39   2015-01-15 19:23:42   
1                1  2015-01-10 20:33:38   2015-01-10 20:53:28   
2                1  2015-01-10 20:33:38   2015-01-10 20:43:41   
3                1  2015-01-10 20:33:39   2015-01-10 20:35:31   
4                1  2015-01-10 20:33:39   2015-01-10 20:52:58   
...            ...                  ...                   ...   
12748981         1  2015-01-10 19:01:44   2015-01-10 19:05:40   
12748982         1  2015-01-10 19:01:44   2015-01-10 19:07:26   
12748983         1  2015-01-10 19:01:44   2015-01-10 19:15:01   
12748984         1  2015-01-10 19:01:44   2015-01-10 19:17:03   
12748985         1  2015-01-10 19:01:45   2015-01-10 19:07:33   

          passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
0                       1           1.59        -73.993896        40.750111   
1                       1           3.30        -74.001648        40.724243   
2                       1           1.80        -73.963341        40.802788   
3                       1           0.50        -74.009087        40.713818   
4                       1           3.00        -73.971176        40.762428   
...                   ...            ...               ...              ...   
12748981                2           1.00        -73.951988        40.786217   
12748982                2           0.80        -73.982742        40.728184   
12748983                1           3.40        -73.979324        40.749550   
12748984                1           1.30        -73.999565        40.738483   
12748985                1           0.70        -73.960350        40.766399   

         store_and_fwd_flag  dropoff_longitude  dropoff_latitude  ...  \
0                         N         -73.974785         40.750618  ...   
1                         N         -73.994415         40.759109  ...   
2                         N         -73.951820         40.824413  ...   
3                         N         -74.004326         40.719986  ...   
4                         N         -74.004181         40.742653  ...   
...                     ...                ...               ...  ...   
12748981                  N         -73.953735         40.775162  ...   
12748982                  N         -73.974976         40.720013  ...   
12748983                  N         -73.969101         40.787800  ...   
12748984                  N         -73.981819         40.737652  ...   
12748985                  N         -73.968643         40.760777  ...   

          fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0                12.0    1.0      0.5        3.25           0.0   
1                14.5    0.5      0.5        2.00           0.0   
2                 9.5    0.5      0.5        0.00           0.0   
3                 3.5    0.5      0.5        0.00           0.0   
4                15.0    0.5      0.5        0.00           0.0   
...               ...    ...      ...         ...           ...   
12748981          5.5    0.0      0.5        1.25           0.0   
12748982          6.0    0.0      0.5        2.00           0.0   
12748983         13.5    0.0      0.5        0.00           0.0   
12748984         10.5    0.0      0.5        2.25           0.0   
12748985          5.5    0.0      0.5        0.00           0.0   

          improvement_surcharge total_amount  PULocationID  DOLocationID  \
0                           0.3        17.05           185           185   
1                           0.3        17.80           212           212   
2                           0.3        10.80           166           166   
3                           0.3         4.80           230           230   
4                           0.3        16.30           162           162   
...                         ...          ...           ...           ...   
12748981                    0.3         7.55           236        

In [26]:
df = cudf.concat([taxi2018, taxi2015])

In [29]:
df.count()

VendorID                 21508860
tpep_pickup_datetime     21508860
tpep_dropoff_datetime    21508860
passenger_count          21508860
trip_distance            21508860
RatecodeID               21508860
store_and_fwd_flag       21508860
PULocationID             21508860
DOLocationID             21508860
payment_type             21508860
fare_amount              21508860
extra                    21508860
mta_tax                  21508860
tip_amount               21508860
tolls_amount             21508860
improvement_surcharge    21508857
total_amount             21508860
pickup_longitude         12748986
pickup_latitude          12748986
RateCodeID               12748986
dropoff_longitude        12748986
dropoff_latitude         12748986
dtype: int64

## Final check
Now to test to see if both years are present as expected

In [28]:
df.query('PULocationID == 2')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  \
1231084          1  2018-01-06 12:56:22   2018-01-06 13:12:52   
1773053          1  2018-01-08 14:58:22   2018-01-08 15:05:58   
1903847          1  2018-01-08 23:13:57   2018-01-08 23:45:34   
5372386          1  2018-01-20 15:02:38   2018-01-20 16:10:51   
261024           1  2015-01-23 21:42:46   2015-01-23 21:46:58   
706444           1  2015-01-11 01:05:10   2015-01-11 01:20:38   
752791           2  2015-01-19 00:13:59   2015-01-19 00:28:32   
1030339          2  2015-01-18 14:37:39   2015-01-18 15:58:39   
2902909          1  2015-01-01 00:34:42   2015-01-01 00:41:22   
3499198          1  2015-01-26 06:16:15   2015-01-26 06:34:06   
3599012          2  2015-01-01 10:49:41   2015-01-01 11:16:53   
3627154          2  2015-01-06 21:06:20   2015-01-06 21:15:19   
4148598          2  2015-01-24 04:25:56   2015-01-24 04:41:15   
5004763          1  2015-01-08 12:46:12   2015-01-08 13:09:41   
5340190          2  2015-01-25 17:25:15   2015-01-25 17:25:50   
5340191          2  2015-01-25 17:25:15   2015-01-25 17:25:50   
5439288          1  2015-01-01 02:08:33   2015-01-01 02:28:12   
5772266          2  2015-01-01 04:34:15   2015-01-01 04:49:04   
5858291          1  2015-01-05 10:35:48   2015-01-05 11:16:31   
6348963          2  2015-01-04 21:35:33   2015-01-04 21:47:54   
6452048          2  2015-01-30 14:29:44   2015-01-30 14:48:53   
8537277          2  2015-01-30 20:46:46   2015-01-30 21:02:16   
8782955          2  2015-01-01 22:40:03   2015-01-01 23:11:33   
9904922          2  2015-01-10 05:25:48   2015-01-10 05:35:56   
10078590         2  2015-01-14 22:21:39   2015-01-14 22:31:19   
10209288         2  2015-01-01 02:34:57   2015-01-01 02:39:02   
10428087         2  2015-01-13 03:30:26   2015-01-13 03:39:56   
10678598         2  2015-01-21 21:07:47   2015-01-21 21:13:46   
10965732         1  2015-01-08 07:46:12   2015-01-08 07:48:52   
11089962         2  2015-01-18 10:21:52   2015-01-18 10:27:24   
11921624         2  2015-01-30 02:20:56   2015-01-30 02:38:22   
11924673         2  2015-01-25 03:37:57   2015-01-25 03:44:23   
12104255         2  2015-01-10 03:52:56   2015-01-10 03:56:31   
12253546         1  2015-01-13 12:44:21   2015-01-13 12:54:53   

          passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
1231084                 1           7.20           1                  N   
1773053                 1           3.20           1                  N   
1903847                 1          12.10           1                  N   
5372386                 1          17.70           2                  N   
261024                  2           1.00           1                  N   
706444                  1           5.70           1                  N   
752791                  1           4.57           1                  N   
1030339                 3          12.53           5                  N   
2902909                 2           2.70           1                  N   
3499198                 1           4.70           1                  N   
3599012                 2          17.43           1                  N   
3627154                 1           2.26           1                  N   
4148598                 1           2.93           1                  N   
5004763                 0          10.90           1                  N   
5340190                 1           0.00           1                  N   
5340191                 1           0.00           1                  N   
5439288                 1           7.70           1                  N   
5772266                 3           6.63           1                  N   
5858291                 1          14.10           1                  N   
6348963                 2           4.44           4                  N   
6452048                 1           3.73           1                  N   
8537277                 1           4.37           1                  N   
8782955  

## Back To Your Workflow
So now you've seen how to use cuSpatial to clean and correlate your spatial data using the NYC taxi data.  You can now perform multi year analytics across the entire range of taxi datasets using your favorite RAPIDS libraries,